numpy/core/src/umath/loops.c.src

```diff
@@ -2618,41 +2618,48 @@ OBJECT_@kind@(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUS
 NPY_NO_EXPORT void
 OBJECT_sign(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSED(func))
 {
-#if defined(NPY_PY3K)
     PyObject *zero = PyLong_FromLong(0);
+
     UNARY_LOOP {
         PyObject *in1 = *(PyObject **)ip1;
         PyObject **out = (PyObject **)op1;
+        PyObject *ret = NULL;
         int v;
-        PyObject *ret;
-        if (PyObject_Cmp(in1 ? in1 : Py_None, zero, &v) == -1) {
-            return;
+
+        if (in1 == NULL) {
+            in1 = Py_None;
+        }
+
+        v = PyObject_RichCompareBool(in1, zero, Py_LT);
+        if (v == 1) {
+            ret = PyLong_FromLong(-1);
+        } else if (v == 0) {
+            v = PyObject_RichCompareBool(in1, zero, Py_GT);
+            if (v == 1) {
+                ret = PyLong_FromLong(1);
+            } else if (v == 0) {
+                v = PyObject_RichCompareBool(in1, zero, Py_EQ);
+                if (v == 1) {
+                    ret = PyLong_FromLong(0);
+                }
+            }
```

Comment:
@jaimefrio I read a bit about NaN. In my opinion, comparing integers with NaN should return `False` as the comparison operators result is a boolean, and the comparison wouldn't be `True`.
Also, in JavaScript, all integer comparisons with NaN evaluates to `false`.
Well, that was my opinion. If it is decided to change the result to `None` as per the community decision. Do keep updated on this one.